sql-plugins/src/main/java/io/cdap/pipeline/sql/plugins/joiner/Joiner.java

```diff
@@ -0,0 +1,277 @@
+  private final JoinerConfig config;
+  private JoinRelType calciteJoinType;
+  private Map<String, String> selectedFields;
+  private List<JoinKey> joinKeys;
+  private Map<String, Integer> stageOrdinalMap;
+
+  public Joiner(JoinerConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public RelBuilder combineInputs(RelBuilder builder, List<String> inputs) {
+    if (inputs.size() != 2) {
+      throw new IllegalArgumentException("Unable to join more than two inputs. Please use multiple joins for more " +
+                                           "than two sources.");
+    }
+
+    // Initialize config
+    init(inputs);
+
+    // Construct the join
+    List<RexNode> conditions = new ArrayList<>();
+    for (JoinKey key: joinKeys) {
+      if (!stageOrdinalMap.containsKey(key.getLeftStage()) || !stageOrdinalMap.containsKey(key.getRightStage())) {
+        throw new IllegalArgumentException("Join key stage is not in the list of input stages.");
+      }
+      int leftStageOrdinal = stageOrdinalMap.get(key.getLeftStage());
+      RexNode leftField = builder.field(inputs.size(), leftStageOrdinal, key.getLeftKey());
+      int rightStageOrdinal = stageOrdinalMap.get(key.getRightStage());
+      RexNode rightField = builder.field(inputs.size(), rightStageOrdinal, key.getRightKey());
+      RexNode condition = builder.call(SqlStdOperatorTable.EQUALS, leftField, rightField);
+      conditions.add(condition);
+    }
+    return builder.join(calciteJoinType, conditions);
+  }
+
+  @Override
+  public RelNode getQuery(RelBuilder builder) {
+    List<RexNode> fields = new ArrayList<>();
+    for (Map.Entry<String, String> selectedFieldsEntry: selectedFields.entrySet()) {
+      String fieldName = selectedFieldsEntry.getKey();
+      String stageName = fieldName.split(Pattern.quote("."))[0];
+      if (!stageOrdinalMap.containsKey(stageName)) {
+        throw new IllegalArgumentException("Selected field stage is not in the list of input stages.");
+      }
+      String alias = selectedFieldsEntry.getValue();
+      RexNode field = builder.alias(builder.field(fieldName), alias);
+      fields.add(field);
+    }
+    return builder.project(fields).build();
+  }
+
+  private void init(List<String> inputs) {
+    selectedFields = new LinkedHashMap<>();
+    joinKeys = new ArrayList<>();
+    // Get a map of the ordinals
+    stageOrdinalMap = new HashMap<>();
+    for (int i = 0; i < inputs.size(); i++) {
+      stageOrdinalMap.put(inputs.get(i), i);
+    }
+
+    // Parse the config strings
+    parseSelectedFields();
+    parseJoinKeys();
+
+    // Parse the join type
+    try {
+      calciteJoinType = JoinRelType.valueOf(config.getJoinType().toUpperCase());
+    } catch (IllegalArgumentException e) {
+      throw new IllegalArgumentException("Unsupported join type: " + config.getJoinType());
+    }
+  }
+
+  /**
+   * Parses the selected fields into a list.
+   */
+  private void parseSelectedFields() {
```

Comment:
this should go in the config class and return the fields instead of modifying state. This encapsulates logic in the correct place, and also makes it more natural to unit test the parsing logic, as you only need to create the config object.

Reply:
Fixed